### DISCUSSION TITLE
fractal stairs pop-up kirigami SVG - generated using Python

### DISCUSSION THREAD

**Discussion by sta-c0000** [amandaghassaei/OrigamiSimulator#43] [Show and tell]
Hello.  I realized from the OrigamiSimulator SVG examples that very little information is required in the files.  I generated a compatible SVG using a simple python3.7+ program without requiring any SVG library (no imports).  Thus, it could be done in any language...

**See the attached generated fractal stairs SVG**, which loads (needs to be unziped first) and runs _very well_ in OrigamiSimulator.

The fractal stairs pop-up kirigami is one of the simplest pop-ups to create by hand: fold paper, using scissors do two cuts half way equally spaced from the edge, invert fold (valley to mountain) of inner piece(s), repeat moving inwards.

This is the very simple python3 program used to generate it:
all released as public domain, do what you wish with it...
```python
#!/usr/bin/python3
'''
Simple fractal stairs kirigami pop-up template generator
outputs: fractalstairs.svg in current dir!
for use with: https://origamisimulator.org/
'''
# paper size in mm
size_x = 210 # A4; US letter = 215.9
size_y = 297 # A4; US letter = 279.4
indent_x = 15 # for each step iteration
levels = 5 # increase paper size for better stability with more levels

boundary = '#000'; mountain = '#f00'; valley = '#00f'; cut = '#0f0'

def line(col, x1, y1, x2, y2):
    return f'<line stroke="{col}" x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}"/>'

def cuboid(level, x1, y1, x2, y2):
    hy = y1 + (y2 - y1) * 0.5
    cl = (line(cut, x1, y1, x1, y2)
        + line(cut, x2, y1, x2, y2)
        + line(mountain, x1, hy, x2, hy)
    # parent's valleys!
        + line(valley, x1-indent_x, hy, x1, hy)
        + line(valley, x2, hy, x2+indent_x, hy))
    if level == levels:
        # last level only: complete valleys
        cl += (line(valley, x1, y1, x2, y1)
             + line(valley, x1, y2, x2, y2))
    return cl

svg = ('<svg xmlns="http://www.w3.org/2000/svg">'
# draw boundary
    + line(boundary, 0, 0, 0, size_y)
    + line(boundary, 0, size_y, size_x, size_y)
    + line(boundary, size_x, size_y, size_x, 0)
    + line(boundary, size_x, 0, 0, 0))

for i in range(levels):
    x1 = indent_x * (i+1)
    x2 = size_x - x1
    frac_y = size_y / 2**(i+2)
    for j in range(2**i):
        y1 = frac_y + frac_y * 4 * j
        y2 = y1 + frac_y * 2
        svg += cuboid(i+1, x1, y1, x2, y2)

svg += '</svg>'

with open('fractalstairs.svg', 'w') as f:
    f.write(svg)
```

[fractalstairs_svg.zip](https://github.com/amandaghassaei/OrigamiSimulator/files/6218024/fractalstairs_svg.zip)

**Comment by amandaghassaei** [comment]
super cool, thanks for posting this (sorry it took me so long to reply).  I'll add it to the example menu.  Do you have a website / name that you want me to include in the credits?

**Comment by sta-c0000** [comment]
It's fun stuff, I don't need attribution, thank you!
